### ISSUE TITLE
Use swissMap for deduping symbols

### ISSUE THREAD

**Issue by cyriltovena** [grafana/pyroscope#2494]
Recently @kolesnikovae  introduces me to SwissMap which seems to be performing better with large map than the runtime one.

See our current ingester flamegraph:

<img width="2254" alt="image" src="https://github.com/grafana/pyroscope/assets/1053421/276b2c0f-dcd6-4f59-a020-5018f1d5252d">

https://flamegraph.com/share/89c22cf6-641f-11ee-b10e-723ba4706d7b

https://github.com/dolthub/swiss

This could be an interesting task for anyone that wants to learn more about how we deduplicate and store symbols.

**Comment by kolesnikovae** [comment]
Yeah, we should give it a try. Also related: https://github.com/golang/go/issues/54766

**Comment by bryanhuhta** [comment]
I did some quick benchmarking with SwissMaps (only using strings) and the results indicated this is a worthwhile venture. The benchmark tested deduplicating a list of strings of varying length (which included duplicates) with increasing lengths of the strings list (N). The implementation code can be viewed [here](https://github.com/grafana/pyroscope/compare/swiss-table).

I compared the to primary APIs of [`deduplicatingMap`](https://github.com/grafana/pyroscope/blob/b4c9552315444ca889f9928bf67006a8c10f8a80/pkg/phlaredb/symdb/dedup_slice.go#L234-L241):

- [`append`](https://github.com/grafana/pyroscope/blob/b4c9552315444ca889f9928bf67006a8c10f8a80/pkg/phlaredb/symdb/dedup_slice.go#L308-L340)
- [`ingest`](https://github.com/grafana/pyroscope/blob/b4c9552315444ca889f9928bf67006a8c10f8a80/pkg/phlaredb/symdb/dedup_slice.go#L256-L306)

**TL;DR** SwissMap outperforms std map in almost all cases, save for with some sizes of relatively small N (< 1000). As N gets larger, SwissMap is the clear winner.

## `append`

When comparing CPU time for `append` with std map and SwissMap, std map will outperform SwissMap at N < ~1000, but as N gets larger than 1000, SwissMap begins to outperform std map by ~25%.

<figure>
  <figcaption>Append CPU time</figcaption>
  <img src="https://github.com/grafana/pyroscope/assets/32787160/e7ee9e72-3f27-4702-8221-9fde8fc9f667">
</figure>

<details>
  <summary>Table View</summary>

|N      |ns/op (std)|ns/op (swiss)| % change<br>smaller is better|
|-------:|-----------:|-------------:|----------------------------------:|
|2      |26         |49           |87                                |
|4      |37         |84           |129                               |
|8      |66         |143          |118                               |
|16     |149        |271          |82                                |
|32     |285        |562          |97                                |
|64     |554        |1053         |90                                |
|128    |1082       |1745         |61                                |
|256    |2157       |3398         |58                                |
|512    |4380       |6707         |53                                |
|1024   |10807      |15214        |41                                |
|2048   |39849      |36219        |-9                                |
|4096   |98661      |82562        |-16                               |
|8192   |224642     |168126       |-25                               |
|16384  |456882     |338324       |-26                               |
|32768  |986451     |701550       |-29                               |
|65536  |1862381    |1467629      |-21                               |
|131072 |3857102    |2819534      |-27                               |
|262144 |7799835    |5699514      |-27                               |
|524288 |15819799   |11503566     |-27                               |
|1048576|31512152   |22718910     |-28                               |
</details>

When comparing the memory performance of `append`, SwissMap is always better than std map by ~30% (when removing outliers).

<figure>
  <figcaption>Append memory allocation</figcaption>
  <img src="https://github.com/grafana/pyroscope/assets/32787160/58769a50-6426-4a6d-930f-cdfa5acd015b">
</figure>

<details>
  <summary>Table View</summary>

|N      |bytes/op (std)|bytes/op (swiss)|% change<br>(smaller is better)|
|-------:|--------------:|----------------:|----------------------------------:|
|2      |0             |0               |N/A                               |
|4      |0             |0               |N/A                               |
|8      |0             |0               |N/A                               |
|16     |0             |0               |N/A                               |
|32     |0             |0               |N/A                               |
|64     |0             |0               |N/A                               |
|128    |0             |0               |N/A                               |
|256    |0             |0               |N/A                               |
|512    |0             |0               |N/A                               |
|1024   |0             |0               |N/A                               |
|2048   |0             |0               |N/A                               |
|4096   |0             |0               |N/A                               |
|8192   |0             |0               |N/A                               |
|16384  |1             |0               |-100                              |
|32768  |2             |1               |-50                               |
|65536  |4             |3               |-25                               |
|131072 |10            |7               |-30                               |
|262144 |20            |14              |-30                               |
|524288 |42            |30              |-29                               |
|1048576|86            |60              |-30                               |

</details>

## `ingest`

When comparing `ingest` CPU time with std map and SwissMap, SwissMap also stays ahead of std map in all cases. Outperforming std map by about ~25%.

<figure>
  <figcaption>Ingest CPU time</figcaption>
  <img src="https://github.com/grafana/pyroscope/assets/32787160/f0e3ec23-d1c5-4ac7-8ce1-7399313a8e22">
</figure>

<details>
  <summary>Table View</summary>

|N      |ns/op (std)|ns/op (swiss)|% change<br>(smaller is better)|
|-------:|--------------:|----------------:|----------------------------------:|
|2      |277.4      |238.3        |-14                               |
|4      |343.4      |286.6        |-17                               |
|8      |445.8      |575.1        |29                                |
|16     |1254       |1083         |-14                               |
|32     |2526       |2043         |-19                               |
|64     |5151       |3949         |-23                               |
|128    |10681      |7698         |-28                               |
|256    |19834      |13999        |-29                               |
|512    |39656      |28176        |-29                               |
|1024   |79786      |59335        |-26                               |
|2048   |173137     |126308       |-27                               |
|4096   |372906     |256878       |-31                               |
|8192   |771281     |523633       |-32                               |
|16384  |1546777    |1092409      |-29                               |
|32768  |3238199    |2249492      |-31                               |
|65536  |6805344    |4863268      |-29                               |
|131072 |13255523   |9484414      |-28                               |
|262144 |26614344   |20347856     |-24                               |
|524288 |58196805   |40653290     |-30                               |
|1048576|126325068  |82757962     |-34                               |

</details>

When comparing the memory performance of `ingest` with std map and SwissMap, SwissMap again produces better performance. Sitting on average ~15% better than std map.

<figure>
  <figcaption>Ingest memory allocation</figcaption>
  <img src="https://github.com/grafana/pyroscope/assets/32787160/4bb6ee11-1ea3-4fd0-aa9b-e8f71b0ab87c">
</figure>

<details>
  <summary>Table View</summary>

|N      |bytes/op (std)|bytes/op (swiss)|% change<br>(smaller is better)|
|-------:|--------------:|----------------:|----------------------------------:|
|2      |304           |280             |-8                                |
|4      |320           |296             |-8                                |
|8      |352           |600             |70                                |
|16     |1307          |1208            |-8                                |
|32     |2672          |2424            |-9                                |
|64     |5766          |4856            |-16                               |
|128    |11716         |9721            |-17                               |
|256    |22982         |19450           |-15                               |
|512    |45514         |38909           |-15                               |
|1024   |90575         |77826           |-14                               |
|2048   |180701        |155661          |-14                               |
|4096   |362057        |311329          |-14                               |
|8192   |719268        |622669          |-13                               |
|16384  |1441412       |1245337         |-14                               |
|32768  |2888889       |2490672         |-14                               |
|65536  |5790633       |4981349         |-14                               |
|131072 |11605813      |9962504         |-14                               |
|262144 |23264225      |19924514        |-14                               |
|524288 |46671056      |39848037        |-15                               |
|1048576|93388601      |79694870        |-15                               |

</details>
